### ISSUE TITLE
What is the main reason for the discrepancy between the inference and detection results of the exported onnx object detection model loaded with openVino C++and opencv and those in Python.

### ISSUE THREAD

**Issue by ldcah** [ultralytics/ultralytics#23345]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

What is the main reason for the discrepancy between the inference and detection results of the exported onnx object detection model loaded with openVino C++and opencv and those in Python.

(1). train and export onnx model
	
	from ultralytics import YOLO
	model = YOLO("yolo26.yaml").load("yolo26n.pt")  
	
	results = model.train(data=r"E:\DataSet\det\data.yaml ", epochs=1000,workers=0)	
	
	results = model.val(workers=0)
	results = model(r"E:\DataSet\det\images")
	results[0].show() 	
	
	success = model.export(format="onnx", opset=12, simplify=True, dynamic=False, imgsz=640)
	
<img width="681" height="502" alt="Image" src="https://github.com/user-attachments/assets/e2b121c9-a476-4943-bbac-b03fa40a99f0" />
<img width="785" height="497" alt="Image" src="https://github.com/user-attachments/assets/a9ce8244-b6bb-47fd-a56b-73043b00a7c3" />

（2.1） python
	from ultralytics import YOLO
	from PIL import Image

	# Load a model
	model = YOLO(r"D:\WORK\AI\ultralytics-8.4.6\runs\detect\train3\weights\best.pt")  # load a custom model

	# Predict with the model
	results = model(r"E:\DataSet\det\images\01.png")  # predict on an image

	# Visualize the results
	for i, r in enumerate(results):
		# Plot results image
		im_bgr = r.plot()  # BGR-order numpy array
		im_rgb = Image.fromarray(im_bgr[..., ::-1])  # RGB-order PIL image

		# Show results to screen (in supported environments)
		r.show()
		# Save results to disk
		r.save(filename=f"results{i}.jpg")

（2.2） C++ code

	struct stRaWHLT
	{
		/// <summary>
		///Magnification ratio
		/// </summary>
		float radioW;
		
		/// <summary>
		///Magnification ratio
		/// </summary>
		float radioH;
		
		/// <summary>
		///Left margin
		/// </summary>
		int left;
		
		/// <summary>
		///Top margin
		/// </summary>
		int top;
	};

	inline Mat LetterBox(const cv::Mat& srcImg, stRaWHLT* _rWH_Left_Top, int dstWidth, int dstHeight, const cv::Scalar& color = cv::Scalar(128, 128, 128)) {
		cv::Size srcShape = srcImg.size();
		Mat dstimg;
		if (srcShape.width == dstWidth && srcShape.height == dstHeight)
		{
			dstimg = srcImg.clone();
			_rWH_Left_Top->radioW = 1;
			_rWH_Left_Top->radioH = 1;
			_rWH_Left_Top->left = 0;
			_rWH_Left_Top->top = 0;
			return dstimg;
		}

		//获取放大比例 -取最小值
		float r = min((float)dstHeight / (float)srcShape.height, (float)dstWidth / (float)srcShape.width);
		int new_un_pad[2] = { (int)std::round((float)srcShape.width * r),(int)std::round((float)srcShape.height * r) };

		auto dw = (float)(dstWidth - new_un_pad[0]);
		auto dh = (float)(dstHeight - new_un_pad[1]);
		dw /= 2.0f;
		dh /= 2.0f;

		int top = int(std::round(dh - 0.1f));
		int bottom = int(std::round(dh + 0.1f));
		int left = int(std::round(dw - 0.1f));
		int right = int(std::round(dw + 0.1f));

		Mat reMat;
		cv::resize(srcImg, reMat, cv::Size(new_un_pad[0], new_un_pad[1]));
		cv::copyMakeBorder(reMat, dstimg, top, bottom, left, right, cv::BORDER_CONSTANT, color);
		reMat.release();
		_rWH_Left_Top->radioW = r;
		_rWH_Left_Top->radioH = r;
		_rWH_Left_Top->left = left;
		_rWH_Left_Top->top = top;
		return dstimg;
	}

	void Detect2026::infer(Mat frame, bool isDrawPred)
	{
		Detect::infer(frame);

		//1.尺寸变换
		Mat dstimg = LetterBox(frame, &_rwh, netWidth, netHeight);

		//2.数据归一化
		Mat blob;
		cv::dnn::blobFromImage(dstimg, blob, 1.0 / 255.0, cv::Size(netWidth, netHeight), cv::Scalar(), true, false);

		//Create tensor from image
		float* input_data = (float*)blob.data;
		ov::Tensor input_tensor = ov::Tensor(model_ptr->input().get_element_type(), model_ptr->input().get_shape(), input_data);
		//  Create an infer request for model inference 
		infer_request.set_input_tensor(input_tensor);
		infer_request.infer();
		// Retrieve inference results 
		ov::Tensor& output_tensor_0 = infer_request.get_output_tensor(0);
		ov::Shape output_shape_0 = output_tensor_0.get_shape();
		float* pdata = output_tensor_0.data<float>();

		//[1,300,6]	 x1,y1,x2,y2,score,class_id
		int rows = output_shape_0[1];
		int cols = output_shape_0[2];

		int index = 0;

		
		for (int i = 0; i < rows; ++i) // 遍历所有的num_pre_boxes
		{
			float max_class_socre = pdata[4];

			if (max_class_socre < 0.05f)
			{
				pdata += cols;
				continue;
			}
			int class_idx = pdata[5];
			if (isFilter(class_idx, max_class_socre))
			{

				float x = (pdata[0] - _rwh.left) / _rwh.radioW;  //x
				float y = (pdata[1] - _rwh.top) / _rwh.radioH;  //y

				float w = (pdata[2]- pdata[0]) / _rwh.radioW;  //w
				float h = (pdata[3]- pdata[1]) / _rwh.radioH;  //h

				int _left = MAX(x, 0);
				int _top = MAX(y, 0);
				int _width = int(w + 0.5);
				int _height = int(h + 0.5);

				stResult.classIds.push_back(class_idx);
				stResult.names.push_back(names[class_idx]);
				stResult.socres.push_back(max_class_socre);
				stResult.bboxes.push_back(Rect(_left, _top, _width, _height));

				//无极大值抑制时增加
				stResult.indices.push_back(index);
				index++;
			}
		
			pdata += cols;
		}
		//无非极大值抑制
		//cv::dnn::NMSBoxes(stResult.bboxes, stResult.socres, nmsScoreThreshold, nmsThreshold, stResult.indices);
		blob.release();
		dstimg.release();
	}

**Comment by UltralyticsAssistant** [comment]
👋 Hello @ldcah, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

This looks like a model export/inference parity question between Python and an ONNX pipeline (OpenVINO C++ + OpenCV) 🧩. This is an automated response, and an Ultralytics engineer will also assist soon ✅

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.
-Please include a tiny sample image (or two), the exact `ultralytics` version, export command, and your full OpenVINO/OpenCV preprocessing + postprocessing steps (including NMS settings and thresholds).  
-If possible, also share the raw ONNX output tensor values for one image from both Python and C++ so we can compare apples-to-apples.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
Padding color should be 114

**Comment by Y-T-G** [comment]
Ultralytics also uses minimum rectangle padding by default:

https://github.com/ultralytics/ultralytics/issues/16388#issuecomment-2363289302

**Comment by ldcah** [comment]
Thank you for your reminder.
I tried using an input image of 640 * 640 for testing, and at this point, the image and padding should not be resized. The results still show differences, and I don't understand where the differences exist.

<img width="577" height="577" alt="Image" src="https://github.com/user-attachments/assets/01238e8e-12f4-44bc-bd11-434ee70126aa" />
![Image](https://github.com/user-attachments/assets/7c88dbc6-7486-4e0e-8dcb-a412fd557c63)

![Image](https://github.com/user-attachments/assets/35717622-1e20-44b0-af31-893286e703f4)

**Comment by glenn-jocher** [comment]
Even with a `640×640` input, mismatched postprocessing is still the most common cause here: by default `model.export(format="onnx", ...)` exports **raw head outputs** (no NMS), so the ONNX output is typically shaped like `(1, 4 + nc, num_preds)` and must be decoded + NMS’d, while your C++ code assumes an NMS-decoded tensor like `[1, N, 6]` (`x1,y1,x2,y2,score,class_id`), which won’t match unless you exported with `nms=True` (or implement the same decode/NMS yourself). Please print and paste the exact OpenVINO output tensor shape you’re getting, and re-export with `nms=True` or follow the expected tensor format described in the [export output tensor notes](https://docs.ultralytics.com/modes/export/) and [ONNX export docs](https://docs.ultralytics.com/integrations/onnx/); also share `yolo checks` output so we can confirm there’s no environment/version mismatch.
